File: java/451.sort-characters-by-frequency.java
/*
 * @lc app=leetcode id=451 lang=java
 *
 * [451] Sort Characters By Frequency
 */
class Solution {

    public String frequencySort(String s) {
        if (s == null) return s;
        int[] freq = new int[256];
        PriorityQueue<Item> queue = new PriorityQueue<>();
        for (int i = 0; i < s.length(); ++i) ++freq[s.charAt(i)];
        for (int i = 0; i < freq.length; ++i) {
            if (freq[i] > 0) queue.add(new Item((char)i, freq[i]));
        }
        char[] res = new char[s.length()];
        for (int i = 0; i < res.length;) {
            Item item = queue.poll();
            for (int j = 0; j < item.f; ++j)
                res[i++] = item.c;
        }
        return new String(res);
    }

    class Item implements Comparable<Item> {
        final char c;
        final int f;
        public Item(char c, int f) {
            this.c = c;
            this.f = f;
        }

        public int compareTo(Item o) {
            if (f > o.f) return -1;
            else if (f < o.f) return 1;
            else return 0;
        }

    }
}

File: go/451.sort-characters-by-frequency.go
/*
 * @lc app=leetcode id=451 lang=golang
 *
 * [451] Sort Characters By Frequency
 */

// @lc code=start
func frequencySort(s string) string {
	freq := make(map[rune]int)
	for _, v := range s {
		if f, b := freq[v]; b {
			freq[v] = f + 1
		} else {
			freq[v] = 1
		}
	}
	maxCnt := 0
	for _, cnt := range freq {
		if cnt > maxCnt {
			maxCnt = cnt
		}
	}
	freqBuckets := make([][]rune, maxCnt)
	for k, cnt := range freq {
		freqBuckets[cnt-1] = append(freqBuckets[cnt-1], k)
	}
	res := make([]rune, len(s))
	pres := 0
	for i := maxCnt - 1; i >= 0; i-- {
		for _, v := range freqBuckets[i] {
			for j := 0; j <= i; j++ {
				res[pres] = v
				pres++
			}
		}
	}
	return string(res)
}

// @lc code=end

